segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/hdfs/HDFSStorage.java

```diff
@@ -196,65 +334,128 @@ public boolean supportsTruncation() {
     }
 
 
-    @Override
-    public int read(SegmentHandle handle, long offset, byte[] buffer, int bufferOffset, int length) throws StreamSegmentException {
-        return call(new ReadOperation(asReadableHandle(handle), offset, buffer, bufferOffset, length, this.context));
-    }
 
     @Override
-    public SegmentProperties getStreamSegmentInfo(String streamSegmentName) throws StreamSegmentException {
-        return call(new GetInfoOperation(streamSegmentName, this.context));
+    public void write(SegmentHandle handle, long offset, InputStream data, int length) throws StreamSegmentException {
+        long traceId = LoggerHelpers.traceEnter(log, "write", handle, offset, length);
+        handle = asWritableHandle(handle);
+        FileStatus lastFile = null;
+        try {
+            lastFile = findStatusForSegment(handle.getSegmentName(), true);
+            if (isSealed(lastFile.getPath())) {
+                throw new StreamSegmentSealedException(handle.getSegmentName());
+            }
+            if (getEpocFromPath(lastFile.getPath()) > this.context.epoch) {
+                throw new StorageNotPrimaryException(handle.getSegmentName());
+            }
+        } catch (IOException e) {
+            HDFSExceptionHelpers.throwException(handle.getSegmentName(), e);
+        }
+
+        Timer timer = new Timer();
+        try (FSDataOutputStream stream = this.context.fileSystem.append(lastFile.getPath())) {
+            if (offset != lastFile.getLen()) {
```

Comment:
We do not have `lastFile` anymore. Let's remove all that `FileDescriptor` and `OperationContext` and make use of what we have. 

Also you will need to update some of these comments, which were tailored for the multi-file approach.

Reply:
Fixed.